subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -825,6 +836,51 @@ private void collectAncestorsSubgraphConstraints(List<EdgeState> incomingEdges)
         ancestorsSubgraphConstraints = SubgraphConstraints.of(constraints);
     }
 
+    private SubgraphConstraints notNull(SubgraphConstraints subgraphConstraints) {
+        return subgraphConstraints == null ? SubgraphConstraints.EMPTY : subgraphConstraints;
+    }
+
+    private SubgraphConstraints getInheritedSubgraphConstraints(EdgeState incomingEdge) {
+        if (incomingEdge.getDependencyState().getDependency().isInheriting()) {
+            return SubgraphConstraints.EMPTY; // skip the inherited constraints of the inheriting node
```

Comment:
I do not understand this comment, nor the way the `if` is branched. Care to explain?

Reply:
I can expand the comment in the code.

This should probably be named `getInheritedSubgraphConstraintsOfParent()` for clarification.

Generally, all nodes need to consider the constraints their parents have inherited. That's the `else` branch.
However, **if I am the node that provides constraints** (e.g. the platform) and I look at the parent that inherited the constraints from me, I don't need to get them. Because they are my `ownSubgraphConstraints`.

Reply:
Thinking about it, maybe we miss an edge case here where we inherit from multiple sources and one of them has itself dependencies (and not only constraints).
Maybe we don't need the `if` path.... I'll investigate.

Thanks for brining this up.

Reply:
No we can not just remove this. Because we always "skip" one level when looking up. I.e. only if our ancestors define a constraint, we can drop our selector. Not if the direct parent defines one.
If we do not keep this special case here, an edge from the "platform" (the node we inherit constraints from) would then think that the parent-of-the-platfrom (the inheriting node) would have defined a constraint and drop its own. But in fact, it has not. Because it is the own constraint, that was inherited.

It is a bit confusing, because "inheriting" does not mean we actually change or move selectors around. We just decide which selectors to "drop".

Reply:
Fixed in b1c0735b0a59beebe1b9649d65dce3ff4c219441